examples/nlp/text_generation_gpt.py

```diff
@@ -0,0 +1,418 @@
+)
+
+"""
+## Load Data
+
+Now, let's download the dataset! The SimpleBooks dataset has a small vocabulary, which
+makes it easier to fit a simple model on.
+"""
+
+keras.utils.get_file(
+    origin="https://dldata-public.s3.us-east-2.amazonaws.com/simplebooks.zip",
+    extract=True,
+)
+dir = os.path.expanduser("~/.keras/datasets/simplebooks/")
+
+# Load simplebooks-92 train set and filter out short lines.
+train_ds = (
+    tf.data.TextLineDataset(dir + "simplebooks-92-raw/train.txt")
+    .filter(lambda x: tf.strings.length(x) > MIN_TRAINING_SEQ_LEN)
+    .batch(BATCH_SIZE)
+    .shuffle(buffer_size=256)
+)
+
+# Load simplebooks-92 validation set and filter out short lines.
+valid_ds = (
+    tf.data.TextLineDataset(dir + "simplebooks-92-raw/valid.txt")
+    .filter(lambda x: tf.strings.length(x) > MIN_TRAINING_SEQ_LEN)
+    .batch(BATCH_SIZE)
+)
+
+"""
+## Tokenize Data
+
+We pre-process the dataset by tokenizing and splitting it into `features` and `labels`.
+"""
+
+
+def preprocess(inputs):
+    outputs = tokenizer(inputs)
+    features = outputs[:, :-1]
+    labels = outputs[:, 1:]
+    return features, labels
+
+
+# Tokenize and split into train and label sequences.
+ds = train_ds.map(preprocess, num_parallel_calls=tf.data.AUTOTUNE).prefetch(
+    tf.data.AUTOTUNE
+)
+ds_valid = valid_ds.map(preprocess, num_parallel_calls=tf.data.AUTOTUNE).prefetch(
+    tf.data.AUTOTUNE
+)
+
+"""
+## Model
+
+We create our scaled down GPT model with the following layers:
+
+- One `keras_nlp.layers.TokenAndPositionEmbedding` layer, which combines the embedding
+for the token and its position. 
+- Multiple `keras_nlp.layers.TransformerDecoder` layers, with the default causal masking.
+The layer has no cross-attention when run with decoder sequence only.
+- One output dense linear layer
+"""
+
+
+def create_model():
+    inputs = keras.layers.Input(shape=(SEQ_LEN,), dtype=tf.int32)
+    # Embedding.
+    embedding_layer = keras_nlp.layers.TokenAndPositionEmbedding(
+        vocabulary_size=VOCAB_SIZE,
+        sequence_length=SEQ_LEN,
+        embedding_dim=EMBED_DIM,
+        mask_zero=True,
+    )
+    x = embedding_layer(inputs)
+    # Transformer decoders.
+    for _ in range(NUM_LAYERS):
+        transformer_block = keras_nlp.layers.TransformerDecoder(
+            num_heads=NUM_HEADS,
+            intermediate_dim=FEED_FORWARD_DIM,
+        )
+        x = transformer_block(x)  # Giving one argument only skips cross-attention.
+    # Output.
+    outputs = keras.layers.Dense(VOCAB_SIZE)(x)
+    model = keras.Model(inputs=inputs, outputs=outputs)
+    loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+    perplexity = keras_nlp.metrics.Perplexity(from_logits=True, mask_token_id=0)
+    model.compile(optimizer="adam", loss=loss_fn, metrics=[perplexity])
+    return model
+
+
+model = create_model()
+
+"""
+Let's take a look at our model summary here! We can see that a large majority of the
```

Comment:
remove the exclamation point here, so the one in the next sentence hits better.

Reply:
edited